### DISCUSSION TITLE
Connecting the UI to a remote daemon

### DISCUSSION THREAD

**Discussion by Dirt-Nasty** [Chia-Network/chia-blockchain#1397] [Q&A]
Has anyone been able to get this to work?  I'm using https://github.com/Chia-Network/chia-blockchain/wiki/Connecting-the-UI-to-a-remote-daemon as a reference written by @dkackman . I'm have the daemon on the remote ports open and I copied the keys to my UI machine but I don't see any of my plots in the farmer tab.  I really like the farmer UI stats and since all my plots are on a remote machine I'd love to see them in the UI.

**Comment by dkackman** [comment]
If you see the URI of the remote daemon in the GUI titlebar and the node status and wallet are updating, then it is working. Managing plots remotely via the GUI hasn't been worked on so the plots don't show up on the farmer tab at this point. 

You should see the harvesters in the "Your Harvester Network" section and the Latest Challenge hashses.

**Reply by Dirt-Nasty** [reply]
I just see one harvester with a hostname of 127.0.0.1 and I do not see the URI of the remote in the titlebar I did fill out the config.yml on the host machine correctly though.  Wouldn't the UI throw some errors if it can't communicate with the daemon?

**Reply by dkackman** [reply]
Have you set the CHIA_ROOT environment variable to point to `~/.chia/mainnet` folder? If it can't find the config file via that environment variable then it doesn't know that you want to deviate from default behavior. I'm sure UX for configuring a remote daemon could be way better but it's kind of an advanced/experimental feature right now so it's rough as far as that goes.

**Reply by Dirt-Nasty** [reply]
do this on the host or GUI machine?

**Reply by dkackman** [reply]
GUI

**Reply by Dirt-Nasty** [reply]
I see the address in the top of the GUI now but just spinning a connecting to wallet.  I'm going to assume the port isn't open.  Do I have to change the host env variable to see the change from localhost to 0.0.0.0 in config.yaml?  I've restarted my processes after that change on the host.  Also when I run sudo ss -tulpn | grep 55400 on the host I get this

`tcp    LISTEN   0        100             127.0.0.1:55400          0.0.0.0:*      users:(("chia_daemon",pid=15503,fd=9)) 
tcp    LISTEN   0        100                 [::1]:55400             [::]:*      users:(("chia_daemon",pid=15503,fd=8))`

**Reply by dkackman** [reply]
when I run that same thing on my remote hose I get 
`tcp    LISTEN  0       100                 0.0.0.0:55400          0.0.0.0:*      users:(("chia_daemon",pid=2925,fd=7))`

The fact that you still have the loopback address implies the daemon isn't picking up that setting change.

Double check - `self_hostname: 0.0.0.0`.

If that looks good do this on the host
`chia stop all -d`
`chia start farmer`

The -d restarts the daemon which is necessary to pick up that change.

**Reply by MilanDeruelle** [reply]
Any additional ways to debug this? I modified the "server", added the config file for the "client" as per gist, I added "CHIA_ROOT=~/.chia/mainnet/" to my .profile and .zshrc (running OS X) but I don't see any changes in my chia GUI application. Not sure where to start looking for the problem... any suggestions?

PS: Running sudo ss -tulpn | grep 55400 on the server gives me 
`tcp     LISTEN   0        100              0.0.0.0:55400          0.0.0.0:*      users:(("chia_daemon",pid=19777,fd=8))`
so I assume, the daemon works fine. However my GUI seems to ignore my config-changes.

**Reply by dkackman** [reply]
Are you 100% sure that on OSX the path to the mainnet folder is `~/.chia/mainnet/` on OSX? I've seen examples where the home directory is embedded down some levels, but I know next to nothing about MACs.

If the URI of your remote machine does not show up in the title bar of the GUI then the problem is locating the mainnet folder. It should look something like this:
![image](https://user-images.githubusercontent.com/5160233/111890456-6ca97f00-89b7-11eb-8f20-a8dc80d0d138.png)

**Reply by MilanDeruelle** [reply]
I've tried two different things now: Using the prebuild dmg image & building the electron app from source.

DMG: I've managed to get the app to get stuck on "Connecting to Wallet" once. Otherwise it seems it has been ignoring the config file. I'm not a 100% sure though, the titlebar looks different on OS X, so I'm not sure if might just not display it in the titlebar. However, the keys I can choose from are the ones from my local config, not the remote daemon, so I'm assuming that's a sign its not working?

Electron from Source: I additionally used the .env file to make sure the path is correctly set. Starting it from the console I get the following output: 
```
> electron
> electron .

/Users/*MyUserName*/Desktop/chia-blockchain/app.asar.unpacked/daemon
false
Error loading config - using defaults
/Users/*MyUserName*/Desktop/chia-blockchain/app.asar.unpacked/daemon
false
/Users/*MyUserName*/Desktop/chia-blockchain/app.asar.unpacked/daemon
false
Running python script
Script ../src/daemon/server.py
child process success
(node:38237) electron: The default of contextIsolation is deprecated and will be changing from false to true in a future release of Electron.  See https://github.com/electron/electron/issues/23506 for more information
No cert
warning, your CHIA_ROOT is set to ~/.chia/mainnet/. Please unset the environment variable and run chia init again
or manually migrate config.yaml
Chia directory /Users/*MyUserName*/.chia/mainnet
/Users/*MyUserName*/.chia/mainnet already exists, no migration action taken
No cert
Have cert

```

**Reply by dkackman** [reply]
This `Error loading config - using defaults` indicates that the gui couldn't find the config file or it has a syntax error. Find the mainnet/config/config.yaml file and run it through something liek https://yamlchecker.com/. Also make sure that whatever you set VHIA_ROOT to resolves to the location of the mainnet folder.

**Comment by dkackman** [comment]
I've added some troubleshooting steps to the how-to: https://github.com/Chia-Network/chia-blockchain/wiki/Connecting-the-UI-to-a-remote-daemon

**Comment by NodeGuy** [comment]
Yes, I got it to work on macOS by running it from the Terminal like this:

```shell
CHIA_ROOT=~/.chia/mainnet /Applications/Chia.app/Contents/MacOS/Chia
```

**Reply by MilanDeruelle** [reply]
Thanks so much, that worked like a charm!

**Reply by dkackman** [reply]
out of curiosity, as I don't have a MAC... does `~` not resolve to `/Users/david/`?

**Reply by NodeGuy** [reply]
Indeed it does; I updated my solution above.

**Reply by winkee01** [reply]
> Yes, I got it to work on macOS by running it from the Terminal like this:
> 
> ```shell
> CHIA_ROOT=~/.chia/mainnet /Applications/Chia.app/Contents/MacOS/Chia
> ```

Can you explain what path that default Chia is using if we don't specify CHIA_ROOT by this way?

**Reply by dkackman** [reply]
Chia by default looks to whatever the OS considers the current home dir for the user + `/.chia/mainnet`. So on linux that's something like `/home/username/.chia/mainnet` and on windows `C:\Users\username\.chia\mainnet`. I'm not a mac user so I'm not sure exactly what that looks like.
